docs/motivation.md

```diff
@@ -0,0 +1,553 @@
+In ML we might want to construct a type such as
+```ocaml
+type s = A of int
+         B of int
+```
+A corresponding Verona type might look like
+```verona
+type S = I32 | I32
+```
+but it is now impossible to discriminate between the two cases. This is however
+easily fixed by defining `S` instead as
+```verona
+class A { var i : I32 }
+class B { var i : I32 }
+
+type S = A | B
+```
+
+#### Explicit nullability
+We can easily encode nullability with a class `None`
+```verona
+type A
+class None {}
+
+f(x : A | None) : S {
+    match x with {
+        (y: A) => { // y : A
+            ...
+        }
+        _ => { // x : None
+            ...
+        }
+    }
+}
+```
+
+### Subtyping and inheritance
+Verona is an object-oriented system with classes and traits. When it comes to
+subtyping, classes are treated nominally, and traits are treated structurally.
+Furthermore there is no subtyping between classes. This means a class
+can be a subtype of a trait, but never a subtype of a class other than itself.
+For example, given the definitions
+```verona
+class C {
+    equal(self: C, other: C): Bool {...}
+}
+
+class D {
+    equal(self: D, other: D): Bool {...}
+}
+
+type Eq = {
+    equal(self: Self, other: Self) : Bool
+}
+
+```
+both `C` and `D` are subtypes of `Eq`, but neither `C` or `D` are subtypes of
+the other.
+
+#### Inheritance
+There are two forms of inheritance in
+Verona. One between classes and one between class and trait.
+
+##### Class inheritance
+Inheritance between classes works as you would expect in a language like
+Java, with the exception that it does not imply subtyping. As an example,
+consider
+```verona
+class C {
+    equal(self: C, other: C): Bool {...}
+    f(...) : ... {...}
+}
+
+// here we reuse functionality of C in D, specifically the method f, but
+// we do not get a subtyping relation D <: C
+class D : C {
+    equal(self: D, other: D): Bool {...}
+    g(...): ... { ... f(...) ... }
+}
+// TODO: find a realistic example
+```
+If we want to describe a relation between `C` and `D`, we can only do so by
+subtyping them under a trait.
+
+There is a case to be made for not allowing inheritance from a
+  class, since this allows unanticipated code reuse. This might however also be
+  a positive.
+
+##### Trait inheritance
+Traits can be used for inheritance with their default methods.
+```verona
+type ToString = {
+    toString(s: Self) : String
+    print(s: Self) : Unit {
+        sys.print(s.toString())
+    }
+}
+
+class C : ToString {
+    toString(s: C) : String {
+        ...
+    }
+}
+// here C will inherit the default implementation of print
+```
+Here, the definition `ToString` contains a default implementation for the print
+method.
+
+* Question: Should traits be allowed to define default fields?
+
+###### External trait inheritance
+A possibility is to allow externally declared trait inheritance, i.e.
+declaration of trait inheritance can be separated from class definition.
+```verona
```

Comment:
I don't understand what this means, or what the example is illustrating. Questions and design discussions are great! Let's move them to a different doc.